cache/cache_test.go

```diff
@@ -398,35 +400,95 @@ func TestValidateWatchRange(t *testing.T) {
 	}
 }
 
-type mockWatcher struct {
-	responses  chan clientv3.WatchResponse
-	registered chan struct{}
-}
+func TestCacheCompactionResync(t *testing.T) {
+	firstSnapshot := &clientv3.GetResponse{
+		Header: &pb.ResponseHeader{Revision: 5},
+		Kvs:    []*mvccpb.KeyValue{{Key: []byte("foo"), ModRevision: 5}},
+	}
+	secondSnapshot := &clientv3.GetResponse{
+		Header: &pb.ResponseHeader{Revision: 20},
+		Kvs:    []*mvccpb.KeyValue{{Key: []byte("foo"), ModRevision: 20}},
+	}
+	fakeClient := &clientv3.Client{
+		Watcher: newMockWatcher(16),
+		KV:      newKVStub(firstSnapshot, secondSnapshot),
+	}
+	cache, err := New(fakeClient, "")
+	if err != nil {
+		t.Fatalf("New cache: %v", err)
+	}
+	defer cache.Close()
+	mw := fakeClient.Watcher.(*mockWatcher)
 
-type kvStub struct{}
+	t.Log("Phase 1: initial getWatch bootstrap")
+	mw.triggerCreatedNotify()
+	<-mw.registered
+	if err = cache.WaitReady(t.Context()); err != nil {
+		t.Fatalf("initial WaitReady: %v", err)
+	}
 
-func (kvStub) Get(ctx context.Context, key string, _ ...clientv3.OpOption) (*clientv3.GetResponse, error) {
-	return &clientv3.GetResponse{Header: &pb.ResponseHeader{Revision: 0}}, nil
-}
+	t.Log("Phase 2: simulate compaction")
+	mw.errorCompacted(10)
 
-func (kvStub) Put(ctx context.Context, key, val string, _ ...clientv3.OpOption) (*clientv3.PutResponse, error) {
-	return nil, nil
-}
+	waitUntil(t, time.Second, 10*time.Millisecond, func() bool { return !cache.Ready() })
+	start := time.Now()
 
-func (kvStub) Delete(ctx context.Context, key string, _ ...clientv3.OpOption) (*clientv3.DeleteResponse, error) {
-	return nil, nil
-}
+	ctxGet, cancelGet := context.WithTimeout(t.Context(), 100*time.Millisecond)
+	defer cancelGet()
+	_, err = cache.Get(ctxGet, "foo", clientv3.WithSerializable())
```

Comment:
Note for the future. If the underlying watch is broken/compacted, we could still serve serializable requests for data we have cache. Dropping is just needed before we start applying new data.